<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Citations &#8212; Brioche 2.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=5ecbeea2" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=27fed22d" />
    <script src="_static/documentation_options.js?v=51b770b3"></script>
    <script src="_static/doctools.js?v=fd6eb6e6"></script>
    <script src="_static/sphinx_highlight.js?v=6ffebe34"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="FAQ" href="faq.html" />
    <link rel="prev" title="Results" href="results.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="citations">
<h1>Citations<a class="headerlink" href="#citations" title="Link to this heading">¶</a></h1>
<p>Below is a list of the key software programs used in the main Brioche and Additional_functions scripts:</p>
<p>magrittr v2.0.4
Bache, S. M., Wickham, H., Henry, L., &amp; Henry, M. L. (2022). Package ‘magrittr.’ R. Package Version.</p>
<p>Picard v3.2.0
Broad Institute. (2019). Picard toolkit. In Broad Institute, GitHub repository [Computer software]. <a class="reference external" href="https://broadinstitute.github.io/picard/">https://broadinstitute.github.io/picard/</a></p>
<p>Blast v2.16.0
Camacho, C., Coulouris, G., Avagyan, V., Ma, N., Papadopoulos, J., Bealer, K., &amp; Madden, T. L. (2009). BLAST+: Architecture and applications. BMC Bioinformatics, 10(1), 421. <a class="reference external" href="https://doi.org/10.1186/1471-2105-10-421">https://doi.org/10.1186/1471-2105-10-421</a></p>
<p>bcftools v1.22
Danecek, P., Bonfield, J. K., Liddle, J., Marshall, J., Ohan, V., Pollard, M. O., Whitwham, A., Keane, T., McCarthy, S. A., &amp; Davies, R. M. (2021). Twelve years of SAMtools and BCFtools. Gigascience, 10(2), giab008.</p>
<p>Nextflow v24.10.2
Di Tommaso, P., Chatzou, M., Floden, E. W., Barja, P. P., Palumbo, E., &amp; Notredame, C. (2017). Nextflow enables reproducible computational workflows. Nature Biotechnology, 35(4), 316–319. <a class="reference external" href="https://doi.org/10.1038/nbt.3820">https://doi.org/10.1038/nbt.3820</a></p>
<p>stringi v1.8.7
Gagolewski, M. (2022). stringi: Fast and Portable Character String Processing in R. Journal of Statistical Software, 103(2), 1–59. <a class="reference external" href="https://doi.org/10.18637/jss.v103.i02">https://doi.org/10.18637/jss.v103.i02</a></p>
<p>entrez-direct v22.4
Kans, J. (2024). Entrez direct: E-utilities on the UNIX command line. In Entrez programming utilities help [Internet]. National Center for Biotechnology Information (US).</p>
<p>Samtools v1.20
Li, H., Handsaker, B., Wysoker, A., Fennell, T., Ruan, J., Homer, N., Marth, G., Abecasis, G., &amp; Durbin, R. (2009). The Sequence Alignment/Map format and SAMtools. Bioinformatics (Oxford, England), 25(16), 2078–2079. <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btp352">https://doi.org/10.1093/bioinformatics/btp352</a></p>
<p>bedtools v2.31.1
Quinlan, A. R., &amp; Hall, I. M. (2010). BEDTools: A flexible suite of utilities for comparing genomic features. Bioinformatics, 26(6), 841–842.</p>
<p>R v4.4.1
R Core Team. (2024). R: A Language and Environment for Statistical Computing [Computer software]. R Foundation for Statistical Computing. <a class="reference external" href="https://www.R-project.org">https://www.R-project.org</a></p>
<p>ggplot2 v4.0.1
Wickham, H. (2011). ggplot2. WIREs Computational Statistics, 3(2), 180–185. <a class="reference external" href="https://doi.org/10.1002/wics.147">https://doi.org/10.1002/wics.147</a></p>
<p>tidyverse v2.0.0
Wickham, H., Averick, M., Bryan, J., Chang, W., McGowan, L., François, R., Grolemund, G., Hayes, A., Henry, L., Hester, J., Kuhn, M., Pedersen, T. L., Miller, E., Bache, S. M., Müller, K., Ooms, J., Robinson, D., Seidel, D. P., Spinu, V., … Woo, K. (2019). Welcome to the Tidyverse. Journal of Open Source Software, 4(43), 1686. <a class="reference external" href="https://doi.org/10.21105/joss.01686">https://doi.org/10.21105/joss.01686</a></p>
<p>dplyr v1.1.4
Wickham, H., Francois, R., Henry, L., Muller, K., &amp; Vaughan, D. (2025). dplyr: A Grammar of Data Manipulation [Computer software]. <a class="reference external" href="https://doi.org/10.32614/CRAN.package.dplyr">https://doi.org/10.32614/CRAN.package.dplyr</a></p>
<p>dbplyr v2.5.1
Wickham, H., Girlich, M., &amp; Ruiz, E. (2025). dbplyr: A “dplyr” Back End for Databases [Computer software]. <a class="reference external" href="https://dbplyr.tidyverse.org/">https://dbplyr.tidyverse.org/</a></p>
<p>readr v2.1.5
Wickham, H., Hester, J., &amp; Bryan, J. (2025). readr: Read Rectangular Text Data [Computer software]. <a class="reference external" href="https://doi.org/10.32614/CRAN.package.readr">https://doi.org/10.32614/CRAN.package.readr</a></p>
<p>tidyr v1.3.1
Wickham, H., &amp; Wickham, M. H. (2017). Package ‘tidyr.’ Easily Tidy Data with’spread’and’gather ()’Functions, 10.</p>
<p>stringr v1.5.2
Wickham, H., &amp; Wickham, M. H. (2019). Package ‘stringr.’ Website: <a class="reference external" href="Http://Stringr">Http://Stringr</a>. Tidyverse. Org, <a class="reference external" href="Https://Github">Https://Github</a>. Com/Tidyverse/Stringr.</p>
<p>data.table v1.15.4
Wiley, M., &amp; Wiley, J. F. (2020). Introduction to data. Table. In Advanced R 4 Data Programming and the Cloud: Using PostgreSQL, AWS, and Shiny (pp. 181–220). Springer.</p>
<p>rentrez v1.2.4
Winter, D. J. (2017). rentrez: An R package for the NCBI eUtils API. The R Journal, 9(2), 520–526. <a class="reference external" href="https://doi.org/10.32614/RJ-2017-058">https://doi.org/10.32614/RJ-2017-058</a></p>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">Brioche</a></h1>






<p>
<iframe src="https://ghbtns.com/github-btn.html?user=plantinformatics&repo=brioche&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>






<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Main</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="setting_up_a_run.html">Setting up a run</a></li>
<li class="toctree-l1"><a class="reference internal" href="results.html">Results</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Citations</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">FAQ</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Use cases</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="usecases/usecase1_remapping.html">Usecase1: Remapping data across reference genomes</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecases/usecase2_insilico_genotyping.html">Usecase2: In silico genotyping of reference genomes</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecases/usecase3_redundancy_accuracy.html">Usecase3: Testing redundancy/accuracy in marker datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecases/usecase4_merge_datasets.html">Usecase4: Merge datasets</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="results.html" title="previous chapter">Results</a></li>
      <li>Next: <a href="faq.html" title="next chapter">FAQ</a></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2026, James O'Dwyer, Gabriel Keeble-Gagnere, Kerrie L Forrest.
      
      |
      <a href="_sources/citations.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/plantinformatics/brioche" class="github">
        <img src="_static/github-banner.svg" alt="Fork me on GitHub" class="github"/>
    </a>
    

    
  </body>
</html>