<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>FAQ &#8212; Brioche 2.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=5ecbeea2" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=27fed22d" />
    <script src="_static/documentation_options.js?v=51b770b3"></script>
    <script src="_static/doctools.js?v=fd6eb6e6"></script>
    <script src="_static/sphinx_highlight.js?v=6ffebe34"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Usecase1: Remapping data across reference genomes" href="usecases/usecase1_remapping.html" />
    <link rel="prev" title="Citations" href="citations.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="faq">
<h1>FAQ<a class="headerlink" href="#faq" title="Link to this heading">¶</a></h1>
<section id="how-many-markers-can-brioche-analyse-in-a-single-analysis">
<h2>1. How many markers can brioche analyse in a single analysis?<a class="headerlink" href="#how-many-markers-can-brioche-analyse-in-a-single-analysis" title="Link to this heading">¶</a></h2>
<p>Brioche is capable of handling large numbers of markers through splitting some of the more compute and memory intensive steps in the pipeline up, but some steps require the holding of complete datasets in memory or reading/writing large files.
These steps can greatly slow down Brioche when too many markers are analysed and greatly increase the required memory to complete. From testing, Brioche starts to significantly slow down once 1 million markers are analysed at once and we recommend
not running anything more than 2 million markers at a time. For larger datasets (large whole genome sequencing) we recommend that the markers are split by chromosome in the input file e.g., if there are 20 chromosomes and 40 million markers, splitting by
chromosome will lead to an average of ~2 million markers per chromosome and per run. Brioche can then be run on each chromosome separately, and then the 1:1 mappings files can be concatenated together for downstream applications.
Other ways to greatly speed up Brioche when working with large numbers of markers include setting stricter blastn settings by passing a -pident and -coverage filter to blastn using the parameter otherblastoptions in the params.config file
For users who have the compute resources to run multiple Brioche runs concurrently a prewritten multijob launch script is avaliable at Additional_functions/Run_brioche_multijob.sh. This can be combined with splitting datasets by chromosome for maximum efficiency and speed</p>
</section>
<section id="what-are-the-best-settings-to-run-brioche-on">
<h2>2. What are the best settings to run Brioche on?<a class="headerlink" href="#what-are-the-best-settings-to-run-brioche-on" title="Link to this heading">¶</a></h2>
<p>The best settings to run Brioche on will depend on both the goal of the project and the type of sequence data and so there are no best settings to run Brioche on. Generally though, if you are trying to run analyses which require exact marker positions e.g., imputation stricter is better.
If you are running datasets generated from data like probe capture, try to set settings which reflect the probe sensitivity e.g., if the probe fails to bind at &lt;95% sequence similarity that should be the identity threshold.
If you are running datasets generated from Reduced representation sequencing lower identity thresholds should be considered to reflect the broad capability of the sequencing methods to sequence broadly similar sequence cut sites.</p>
</section>
<section id="the-dart-scripts-don-t-seem-to-work">
<h2>3. The DArT scripts don’t seem to work?<a class="headerlink" href="#the-dart-scripts-don-t-seem-to-work" title="Link to this heading">¶</a></h2>
<p>The DArT scripts to either create the Brioche input targets file or anchor the genotypes to new genomes are designed to only work with 1-row format DArT reports. This means that if the report you have is in 2-row format the scripts won’t work.
If your format is 1-row and the scripts aren’t working, let us know, there is a lot of variability in what is provided in a DArT report format so we might need to update the scripts to work against different formats we didn’t see during testing.</p>
</section>
<section id="some-analysis-steps-seem-to-be-taking-a-very-long-time-or-are-restarting-why-is-that">
<h2>4. Some analysis steps seem to be taking a very long time or are restarting why is that?<a class="headerlink" href="#some-analysis-steps-seem-to-be-taking-a-very-long-time-or-are-restarting-why-is-that" title="Link to this heading">¶</a></h2>
<p>If some steps are taking a really long time to run or a restarting it might be an indication that not enough memory is allocated to the steps. The memory settings for different steps can be changed in the file nextflow.config in the main Brioche directory.
Try increasing the memory for the section which is withLabel: ‘large’ or the section withLabel ‘merge’. If this doesn’t help or you do not have access to more compute resources, you might need to run brioche in subsets e.g., by splitting your markers by chromosome</p>
</section>
<section id="brioche-seems-to-crash-and-not-install-the-conda-environments">
<h2>5. Brioche seems to crash and not install the conda environments<a class="headerlink" href="#brioche-seems-to-crash-and-not-install-the-conda-environments" title="Link to this heading">¶</a></h2>
<p>Some HPC systems require the user to download/install programs in a special download/install partition of the HPC. If this is the case, Brioche may fail is just run normally. We recommend that you launch brioche (e.g., the example_run_script_brioche.sh script) with your install/download partition selected.
For the additional functions which require the conda environment brioche-vcf being built (e.g., anchoring genotypes, or converting DArT/VCF formats to brioche input) install the conda environment first by running the script build_brioche-vcf_env.sh in the Additional_functions folder on the download/install partition of your HPC</p>
</section>
<section id="i-don-t-have-access-to-many-modules-and-can-t-load-nextflow-r-etc">
<h2>6. I don’t have access to many modules and can’t load Nextflow/R etc<a class="headerlink" href="#i-don-t-have-access-to-many-modules-and-can-t-load-nextflow-r-etc" title="Link to this heading">¶</a></h2>
<p>If you don’t have access to any of the preinstalled modules we recommend you begin by trying to install Miniconda following the instructions outlined in the installation section of this manual. After that, you can try and install R, Nextflow and Git into your
conda base environment using conda install. From there, you will need to add in a few extra lines to each brioche launch script to make sure conda is being called before you begin e.g.,</p>
<blockquote>
<div><div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="nb">eval</span><span class="w"> </span><span class="s2">&quot;</span><span class="k">$(</span>conda<span class="w"> </span>shell.bash<span class="w"> </span>hook<span class="k">)</span><span class="s2">&quot;</span>

conda<span class="w"> </span>activate<span class="w"> </span>base
</pre></div>
</div>
</div></blockquote>
</section>
<section id="my-hpc-does-not-use-slurm-it-uses-pbs-qsub-lsf-sge-etc">
<h2>7. My HPC does not use slurm it uses PBS/QSUB/LSF/SGE etc.<a class="headerlink" href="#my-hpc-does-not-use-slurm-it-uses-pbs-qsub-lsf-sge-etc" title="Link to this heading">¶</a></h2>
<p>Currently Brioche is written to work in a slurm management system but it may be possible for an experienced user to change the required settings to match the requirements of different HPC management systems.
We have not tested this but feel free to try and update the ##SBATCH commands and the settings in the nextflow.config file to reflect your HPC management system. Feel free to let us know how it works out. If there is enough demand we
might look into ways to make Brioche more HPC management system agnostic.</p>
</section>
<section id="brioche-crashed-and-nothing-above-helps">
<h2>8. Brioche crashed and nothing above helps<a class="headerlink" href="#brioche-crashed-and-nothing-above-helps" title="Link to this heading">¶</a></h2>
<p>We are still actively testing and building on Brioche. If you experience any unusual behaviour or crashes feel free to reach out or create an issue in the github. We appreciate the feedback and any fixed bug is a step towards Brioche working perfectly.</p>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">Brioche</a></h1>






<p>
<iframe src="https://ghbtns.com/github-btn.html?user=plantinformatics&repo=brioche&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>






<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Main</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="setting_up_a_run.html">Setting up a run</a></li>
<li class="toctree-l1"><a class="reference internal" href="results.html">Results</a></li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">FAQ</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#how-many-markers-can-brioche-analyse-in-a-single-analysis">1. How many markers can brioche analyse in a single analysis?</a></li>
<li class="toctree-l2"><a class="reference internal" href="#what-are-the-best-settings-to-run-brioche-on">2. What are the best settings to run Brioche on?</a></li>
<li class="toctree-l2"><a class="reference internal" href="#the-dart-scripts-don-t-seem-to-work">3. The DArT scripts don’t seem to work?</a></li>
<li class="toctree-l2"><a class="reference internal" href="#some-analysis-steps-seem-to-be-taking-a-very-long-time-or-are-restarting-why-is-that">4. Some analysis steps seem to be taking a very long time or are restarting why is that?</a></li>
<li class="toctree-l2"><a class="reference internal" href="#brioche-seems-to-crash-and-not-install-the-conda-environments">5. Brioche seems to crash and not install the conda environments</a></li>
<li class="toctree-l2"><a class="reference internal" href="#i-don-t-have-access-to-many-modules-and-can-t-load-nextflow-r-etc">6. I don’t have access to many modules and can’t load Nextflow/R etc</a></li>
<li class="toctree-l2"><a class="reference internal" href="#my-hpc-does-not-use-slurm-it-uses-pbs-qsub-lsf-sge-etc">7. My HPC does not use slurm it uses PBS/QSUB/LSF/SGE etc.</a></li>
<li class="toctree-l2"><a class="reference internal" href="#brioche-crashed-and-nothing-above-helps">8. Brioche crashed and nothing above helps</a></li>
</ul>
</li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Use cases</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="usecases/usecase1_remapping.html">Usecase1: Remapping data across reference genomes</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecases/usecase2_insilico_genotyping.html">Usecase2: In silico genotyping of reference genomes</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecases/usecase3_redundancy_accuracy.html">Usecase3: Testing redundancy/accuracy in marker datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecases/usecase4_merge_datasets.html">Usecase4: Merge datasets</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="citations.html" title="previous chapter">Citations</a></li>
      <li>Next: <a href="usecases/usecase1_remapping.html" title="next chapter">Usecase1: Remapping data across reference genomes</a></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2026, James O'Dwyer, Gabriel Keeble-Gagnere, Kerrie L Forrest.
      
      |
      <a href="_sources/faq.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/plantinformatics/brioche" class="github">
        <img src="_static/github-banner.svg" alt="Fork me on GitHub" class="github"/>
    </a>
    

    
  </body>
</html>