<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Installation &#8212; Brioche 2.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=5ecbeea2" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=27fed22d" />
    <script src="_static/documentation_options.js?v=51b770b3"></script>
    <script src="_static/doctools.js?v=fd6eb6e6"></script>
    <script src="_static/sphinx_highlight.js?v=6ffebe34"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Setting up a run" href="setting_up_a_run.html" />
    <link rel="prev" title="#Introduction" href="introduction.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="installation">
<h1>Installation<a class="headerlink" href="#installation" title="Link to this heading">¶</a></h1>
<section id="downloading-brioche">
<h2>Downloading brioche:<a class="headerlink" href="#downloading-brioche" title="Link to this heading">¶</a></h2>
<p>Brioche can be downloaded directly from the github page as shown below.</p>
<p>.. code-block:: console</p>
<p>git clone https://github.com/plantinformatics/brioche.git</p>
<p>after downloading you can navigate to the brioche folder.</p>
<p>.. code-block:: console</p>
<p>cd brioche</p>
<p>Here you will have a folder structure laid out like the following:</p>
<p>.. image:: Images/Brioche_github_downloaded.png
:alt: Brioche downloaded from github
:width: 300px
:align: center</p>
</section>
<section id="software-dependencies">
<h2>Software dependencies:<a class="headerlink" href="#software-dependencies" title="Link to this heading">¶</a></h2>
<p>Brioche uses Nextflow as a workflow manager. Brioche also uses conda as a package manager and git for recording relevant metadata. Each of these are required to be installed either as modules on the HPC system
or directly by the user.</p>
<p>e.g.,</p>
<p>.. code-block:: console</p>
<p>module load Nextflow</p>
<p>or to download nextflow</p>
<p>.. code-block:: console</p>
<p>wget -qO- https://get.nextflow.io | bash</p>
<p>The same applies for conda</p>
<p>.. code-block:: console</p>
<p>module load Miniconda3</p>
<p>or to download conda we navigate to the desired location for storing conda (usually your home directory), download,unpack and initialise conda</p>
<p>.. code-block:: console</p>
<p>wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh -O /path/to/desired/directory/miniconda/miniconda.sh
bash /path/to/desired/directory/miniconda/miniconda.sh -b -u -p /path/to/desired/directory/miniconda/
/path/to/desired/directory/miniconda/bin/conda init bash</p>
<p>Finally, for git</p>
<p>.. code-block:: console</p>
<p>module load git</p>
<p>or (asssuming Debian/Ubuntu Linux)</p>
<p>.. code-block:: console</p>
<p>apt-get install git</p>
<p>Brioche is currently written to work within <strong>slurm</strong> HPC systems so will not run out of the box on other systems.
If the user is interested in working with other HPC management systems individual .sh files will need to have their headers updated and the nextflow.config parameters updated manually</p>
<p>During the first run of Brioche, the main conda environment will be built automatically from the conda_environment.yml file provided in the brioche directory.
NOTE: this means that the launch partition of Brioche (e.g., example_run_script_brioche.sh) needs to have access to the internet so set the partition to one which can download.</p>
<p>Some additional software is required for extended functionality in Brioche. This can be found in the Additional_functions folder and will be installed as a conda environment using the file brioche-vcf.yaml
This software environment is used to do additional insilico and other functionality. This will install when running these scripts, or you can install it manually by running build_brioche-vcf_env.sh</p>
</section>
<section id="file-layout">
<h2>File layout:<a class="headerlink" href="#file-layout" title="Link to this heading">¶</a></h2>
<p>Brioche comes with a number of different files and folders but only a small number are important for the user to be aware of.</p>
<p>e.g., the following folders are usefull for the user</p>
<p>example_files/</p>
<p>example_files contains example formatted files for each optional and mandatory input and parameter file Brioche can use</p>
<p>.. image:: Images/example_files_folder.png
:alt: example_files folder
:width: 300px
:align: center</p>
<p>Additional_functions/</p>
<p>Additional_functions contains a collection of helpful bash and R scripts designed to expand the regular use of Brioche through 1. helping create input files from diverse datatypes (e.g.,run_convert_vcf_to_brioche_input.sh)
and 2. Running specialised Brioche runs including testing the parameter space for marker uniqueness (e.g., Run_brioche_optimise.sh), or insilico adding reference genome genotypes (e.g., Run_insilico_add.sh)</p>
<p>Each of these scripts contains a short description of how to run in the file and what settings to change but additional details can be found in the relevant usecase described in this User guide.</p>
<p>.. image:: Images/Additional_functions.png
:alt: Additional_functions folder
:width: 300px
:align: center</p>
<p>The following files are useful for the user</p>
<p>params.config</p>
<p>This file contains a list of all parameters that can be changed in Brioche allowing for greater control of a run</p>
<p>nextflow.config</p>
<p>This file contains additional details about resource allocation for the different types of runs Brioche can do. This can be edited to allocated great/less memory, CPUs etc per individual run if extremely large/small datasets
and reference genomes are being used and standard allocations are not efficient.</p>
<p>example_run_script_brioche.sh</p>
<p>This file contains a simple run script for how to run Brioche. It can be opened, the slurm settings updated and then launched with sbatch commands.</p>
<p>Introduction:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>
To return to the Introduction page go to {doc}`Introduction &lt;Introduction&gt;`


Run setup:
~~~~~~~~~~

For setting up a run go to {doc}`Setting up a run &lt;setting_up_a_run&gt;`

</pre></div>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">Brioche</a></h1>






<p>
<iframe src="https://ghbtns.com/github-btn.html?user=plantinformatics&repo=brioche&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>






<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Main</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">#Introduction</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#downloading-brioche">Downloading brioche:</a></li>
<li class="toctree-l2"><a class="reference internal" href="#software-dependencies">Software dependencies:</a></li>
<li class="toctree-l2"><a class="reference internal" href="#file-layout">File layout:</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="setting_up_a_run.html">Setting up a run</a></li>
<li class="toctree-l1"><a class="reference internal" href="results.html">Results</a></li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">FAQ</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Use cases</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="usecases/usecase1_remapping.html">Usecase1: Remapping data across reference genomes</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecases/usecase2_insilico_genotyping.html">Usecase2: In silico genotyping of reference genomes</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecases/usecase3_redundancy_accuracy.html">Usecase3: Testing redundancy/accuracy in marker datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecases/usecase4_merge_datasets.html">Usecase4: Merge datasets</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="introduction.html" title="previous chapter">#Introduction</a></li>
      <li>Next: <a href="setting_up_a_run.html" title="next chapter">Setting up a run</a></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2026, James O'Dwyer, Gabriel Keeble-Gagnere, Kerrie L Forrest.
      
      |
      <a href="_sources/installation.md.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/plantinformatics/brioche" class="github">
        <img src="_static/github-banner.svg" alt="Fork me on GitHub" class="github"/>
    </a>
    

    
  </body>
</html>