<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Setting up a run &#8212; Brioche 2.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=5ecbeea2" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=27fed22d" />
    <script src="_static/documentation_options.js?v=51b770b3"></script>
    <script src="_static/doctools.js?v=fd6eb6e6"></script>
    <script src="_static/sphinx_highlight.js?v=6ffebe34"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Results" href="results.html" />
    <link rel="prev" title="Installation" href="installation.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="setting-up-a-run">
<h1>Setting up a run<a class="headerlink" href="#setting-up-a-run" title="Link to this heading">¶</a></h1>
<section id="brioche-inputs">
<h2>Brioche inputs:<a class="headerlink" href="#brioche-inputs" title="Link to this heading">¶</a></h2>
<p>Brioche has one required input file (the brioche targets file) which is formatted as a tsv file with the following columns.
1. ID  (the name of each marker)
2. Sequence (the sequence to test for uniqueness and mapping in Brioche with the target SNP position replaced with a set not ACGT character e.g., D or X
3. Target.bp (the position of the SNP site of interest in the sequence provided
4. Target.base (a description of the REF/ALT of the expected SNP in the format [REF/ALT] for the marker)</p>
<a class="reference internal image-reference" href="_images/brioche_targets_file.png"><img alt="Targets file" class="align-center" src="_images/brioche_targets_file.png" style="width: 300px;" />
</a>
<p>This file can be created manually by the user, or some of the scripts in the Additional_funtions/ folder may assist</p>
<p>e.g., If you have a VCF file and want to extract sequences from it to run Brioche you can run the file run_convert_vcf_to_brioche_input.sh e.g.,
vcfin= Your vcf file
refgenome= The reference genome you want to extract flanking sequence from
fragmentsize= The size of the fragment around the SNP to return e.g., 75bp dowsntream and 75bp upstream of SNP is 150
chrommapping= If the VCF has different chromosome names than tha reference, this file contains the pairwise comparison on names</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">cd Additional_funtions/</span>
<span class="go">bash run_convert_vcf_to_brioche_input.sh \</span>
<span class="go">--vcfin file.vcf.gz \</span>
<span class="go">--refgenome genome1.fa \</span>
<span class="go">--fragmentsize 150 \</span>
<span class="go">--chrommapping chrom_mapping.tsv \</span>
<span class="go">--outdir ./brioche_out</span>
</pre></div>
</div>
<p>Or if you have a DArT report (1 row format). This requires R +v4.0.0 to either be installed as a module or for the user to install through conda. It also requires the installation of the R package “stringi” which occurs inside the script
Note: this will extract the AlleleSequence from the DArT report, the TrimmedSequence can be used instead if preferred by changing 2 lines in the script (Ln 93,101 See Usage documentation inside the script)
dartfile= the path and name of the 1 row format dart report file being read in
targetmarker= the letter to replace the SNP site with (e.g., D or X or N)</p>
<p>e.g.,</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">cd Additional_funtions/</span>
<span class="go">module load R4.4.0</span>
<span class="go">Rscript DArT2Brioche_convert.R \</span>
<span class="go">--dartfile &quot;dart_testing_build_briochefiles.csv&quot; \</span>
<span class="go">--targetmarker &quot;D&quot;</span>
</pre></div>
</div>
</section>
<section id="parameters-file">
<h2>Parameters file:<a class="headerlink" href="#parameters-file" title="Link to this heading">¶</a></h2>
<p>The parameters file for each run of brioche is found in the file params.config.</p>
<p>This file can be found in the brioche/ directory and has been populated with a number of default settings allowing for quick running of data.</p>
<p>This file can be renamed so long as the variable –paramfile in the nextflow run (e.g., example_run_script_brioche.sh) is updated to the new name; otherwise the run will default to params.config.</p>
<p>There are 41 different parameter settings that can be changed in Brioche with a description of what each does and what each can be set to within the parameters file</p>
<a class="reference internal image-reference" href="_images/paramsfile.png"><img alt="Parameters file" class="align-center" src="_images/paramsfile.png" style="width: 300px;" />
</a>
<p>These parameters can broadly be categorised into 6 groups of variables with some of the main listed below</p>
<section id="run-execution-settings">
<h3>1. Run execution settings<a class="headerlink" href="#run-execution-settings" title="Link to this heading">¶</a></h3>
<p>mode (whether to run as test or prod for real or test data)
notifications (whether to receive an email upon completion)
resultsdir (where to save results)</p>
</section>
<section id="slurm-settings">
<h3>2. slurm settings<a class="headerlink" href="#slurm-settings" title="Link to this heading">¶</a></h3>
<p>hpcaccount (the account name sbatch jobs should have to run)
shortqueue (the queue/partition for jobs which take &lt;12 hours)
longqueue (the queue/partition for jobs which take &gt;12 hours)
queuesize (the maximum number of slurmjobs to submit at any given time)</p>
</section>
<section id="blastn-settings">
<h3>3. BLASTn settings<a class="headerlink" href="#blastn-settings" title="Link to this heading">¶</a></h3>
<p>evalue (minimum evalue to report BLASTn hits)
dust (whether to filter highly redundant sequences and not return any results e.g., long strings of CGCG or Ns)
otherblastoptions (any other BLASTn specific parameter to use e.g., –wordsize)
querychuncksize (Maximum number of markers parsed to BLASTn. If increased, memory requirements increase but can lead to small increases in speed)</p>
</section>
<section id="filtering-options">
<h3>4. Filtering options<a class="headerlink" href="#filtering-options" title="Link to this heading">¶</a></h3>
<p>coverage (the minimum coverage required to pass the coverage filter)
pident (the minimum pairwise identity between the marker sequence and reference genome segment to pass the identity filter)
maximumhits (the maximum number of BLASTn hits a marker can have before the marker is filtered out
istarget3primeend (whether to filter sequences which have mutations at the 3’ end of the sequence, relevant for probe and SNP chip design data)
extendablebps (the number of bases required to be identical from 3’ if istarget3primend is set to TRUE)
keeplocalduppos (whether to treat local duplications within a small range of the genome as unique and map to the first copy or redundant and filter [will filter if the SNP site differs regardless of what this is set to])
localdupdist (the distance upstream and downstream in base pairs to search for local duplications if keeplocalduppos is set to ‘yes’)</p>
</section>
<section id="anaylsis-scope-restricting-settings">
<h3>5. Anaylsis scope restricting settings<a class="headerlink" href="#anaylsis-scope-restricting-settings" title="Link to this heading">¶</a></h3>
<p>restrict2chrom (only analyse hits against this chromosome of the reference genome)
chromstoexclude (a list of chromosomes to exclude from the reference genome)</p>
</section>
<section id="providing-supplemental-biological-information-to-assist-with-determining-optimal-marker-positions">
<h3>6. Providing supplemental biological information to assist with determining optimal marker positions<a class="headerlink" href="#providing-supplemental-biological-information-to-assist-with-determining-optimal-marker-positions" title="Link to this heading">¶</a></h3>
<p>usetargetchrom (whether to use a pairwise list of preference chromosomes to match each marker to. Will result in equal quality hits across multiple chromosomes being preferenced to target chromosome and uniquely positioned)
markertargetsites (two column csv which lists each marker and the chromosome to map to [chromosomes can be in the old reference genome chromosome name], see example_files/targetchromsknownmarkers.csv)
chromchrommatch (a two column csv listing the names of the original reference genome chromosomes and the equivalent chromosome name for the new genome, see example_files/Chromesome-chromosome_mappings.csv)</p>
<p>usesharedmarkersmap (whether to use a csv list of closely linked marker names to inform the relative position of each marker. Will result in equal quality hits across multiple chromosomes/within a chromosome being preferenced to target a position closest in bp on the reference genome to the given linked markers and causing this marker to report as uniquely positioned)
similarmarkersmap (a multicolumn csv file which lists the target marker in the leftmost column and closely linked markers in the following columns, see example_files/linkedSNPs.csv)</p>
<p>useldedgemap (whether to use a raw linkage disequilibrium table between markers to match the optimal marker position on the genome. Will result in equal quality hits across/within chromosomes being preferenced to target the position which has the lowest bp distance from the most closely linked markers and cause this marker to report as uniquely positioned)
ldedgemap (the path to a pairwise linkage disequilibrium output table [format PLINK, TASSEL, HAPLOVIEW long tables] for linkage disequilibrium scores, see example_files/examplelinkagedisequalibriumfile.ld )</p>
<p>usegeneticmap (whether to use a genetic mapping file to infirm on the relative position of each marker. Will result in equal quality hits across multiple chromosomes being preferenced to target the identified chromosome by genetic mapping. Will also result in markers which map across multiple chromosomes in genetic mapping to be marked as not unique if any hits to those secondary chromosomes and which passed identity/coverage filtering  were found from BLASTn)
geneticmap (the path to a multicolumn csv with the first column being the marker name, the second the number of chromosomes this marker mapped to, the next 5 being the chromosome names it mapped to, and the next 5 columns being the counts which each chromosome was mapped to for the marker, see example_files/genetic_mapping_input.csv. Note, also requires chromchrommatch to be present to convert chrom names between old and new references)</p>
</section>
</section>
<section id="run-parameters-and-goals">
<h2>Run parameters and goals<a class="headerlink" href="#run-parameters-and-goals" title="Link to this heading">¶</a></h2>
<p>The run configuration you choose depends on your workflow goal. We provide a more detailed breakdown of how some settings can be changed to best run for different usecases below but in general the settings desired should reflect</p>
<section id="the-nature-of-your-data-and-the-reference">
<h3>1. The nature of your data and the reference<a class="headerlink" href="#the-nature-of-your-data-and-the-reference" title="Link to this heading">¶</a></h3>
<p>e.g., If your data is a probe design that is highly specific then a high coverage and identity is best to reflect the high specificty expected in your data.</p>
<p>e.g., If you know that the reference sequence you are mapping to is fairly diverged coverage and identity should be slightly lower to reflect the expected divergence outside of the SNP site.</p>
<p>e.g., If your data is generated from more redundancy based methods e.g., reduced represenation sequencing there will likely be a very large number of markers that are not fully unique and so settings should reflect this fact</p>
</section>
<section id="the-purpose-of-the-newly-anchored-data">
<h3>2. The purpose of the newly anchored data<a class="headerlink" href="#the-purpose-of-the-newly-anchored-data" title="Link to this heading">¶</a></h3>
<p>e.g., If the intention is to identify strictly mapping data for position sensitive analyses (e.g., imputation) then settings should be conservative as it is more important to remove any potentially non uniquely mapped markers</p>
</section>
<section id="settings-are-not-linear-conservative-liberal">
<h3>3. Settings are not linear conservative/liberal<a class="headerlink" href="#settings-are-not-linear-conservative-liberal" title="Link to this heading">¶</a></h3>
<p>Lastly, consider that certain settings may result in more/fewer markers being mapped uniquely but this relationship is not necessarily linear. e</p>
<p>e.g., while increasing the pident and coverage thresholds increases the number of markers which failed at this early filtering step, it might also remove poorly mapping hits for markers resulting in only unique hits remaining and increasing the final
number of uniquely mapping markers.
This is why either setting setting to reflect the nature of the data (point 1) or exploring the results of multiple runs is recommended (see Additional_functions/Run_brioche_optimise.sh)</p>
<p>Other specific goals and scenarios can be seen in the example of different usecases.</p>
<ul class="simple">
<li><p><a class="reference internal" href="usecases/usecase1_remapping.html"><span class="doc">Usecase1: Remapping data across reference genomes</span></a></p></li>
<li><p><a class="reference internal" href="usecases/usecase2_insilico_genotyping.html"><span class="doc">Usecase2: In silico genotyping of reference genomes</span></a></p></li>
<li><p><a class="reference internal" href="usecases/usecase3_redundancy_accuracy.html"><span class="doc">Usecase3: Testing redundancy/accuracy in marker datasets</span></a></p></li>
<li><p><a class="reference internal" href="usecases/usecase4_merge_datasets.html"><span class="doc">Usecase5: Merge datasets</span></a></p></li>
</ul>
</section>
</section>
<section id="simple-runs">
<h2>Simple runs<a class="headerlink" href="#simple-runs" title="Link to this heading">¶</a></h2>
<p>Once parameters are set and the targets file (Brioche input file) is created running Brioche is very straight forward.</p>
<p>In the brioche landing directory there is an example .sh file for running Brioche, example_run_script_brioche.sh</p>
<p>In this file you will need to</p>
<ol class="arabic simple">
<li><p>Update the slurm sbatch commands</p></li>
</ol>
<p>“””</p>
<p># Multithreaded (SMP) job: must run on one node
#SBATCH –nodes=1</p>
<p># The name of the job:
#SBATCH –job-name=”example_slurmjob”</p>
<p># The project ID which this job should run under:
#SBATCH –account=”default”</p>
<p># The partition in which to submit the job:
#SBATCH –partition=”batch”</p>
<p># Maximum number of tasks/CPU cores used by the job:
#SBATCH –ntasks=1
#SBATCH –cpus-per-task=8</p>
<p># The total amount of memory in megabytes in the job:
#SBATCH –mem=40GB</p>
<p># Send yourself an email when the job:
# aborts abnormally (fails)
#SBATCH –mail-type=FAIL</p>
<p># Use this email address:
#SBATCH <a class="reference external" href="mailto:--mail-user=user123&#37;&#52;&#48;mail&#46;mail">–mail-user=user123<span>&#64;</span>mail<span>&#46;</span>mail</a></p>
<p># The maximum running time of the job in days-hours:mins:sec
#SBATCH –time=24:0:00</p>
<p>“””</p>
<ol class="arabic simple" start="2">
<li><p>update some settings</p></li>
</ol>
<p>Paying special attention to the target file (brioche targets file), the file paths needed,
and making sure the three programs Nextflow, conda, and git are loaded in however you installed them e.g., through module load or source activate.</p>
<p>“””
#Build target table for the new brioche format
target=/filepath/briochetargetdesign.tsv</p>
<p># Used in naming of results files
chipname=”probename”</p>
<p># directory of reference genome
genomedir=/filepathgenome/</p>
<p># Genome name without extension
genome=”genomename”</p>
<p># genome name with
fasta=”genomename.fa”</p>
<p># param file can be set as anything but if not set in the nextflow run command below will default to the params.config present in the same directory as the main.nf
params=”/filepath/brioche/params.config”
briochemain=”/filepath/brioche/main.nf”</p>
<p># required modules
module load Nextflow
module load Miniconda3
module load git</p>
<blockquote>
<div><p>workdir=”/filepath/brioche/results/${genome}_${chipname}/”
resultsdir=”/filepath/brioche/results/${genome}_${chipname}/”
mkdir -p ${resultsdir}
mkdir -p ${workdir}
cd ${workdir}</p>
</div></blockquote>
<p>“””</p>
<p>After updating the script can be run from a terminal with</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">sbatch example_run_script_brioche.sh</span>
</pre></div>
</div>
<section id="introduction">
<h3>Introduction:<a class="headerlink" href="#introduction" title="Link to this heading">¶</a></h3>
<p>To return to the Installation page go to <a class="reference internal" href="installation.html"><span class="doc">Installation</span></a></p>
</section>
<section id="results">
<h3>Results:<a class="headerlink" href="#results" title="Link to this heading">¶</a></h3>
<p>For investigating the results of a run go to <a class="reference internal" href="results.html"><span class="doc">Results</span></a></p>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">Brioche</a></h1>






<p>
<iframe src="https://ghbtns.com/github-btn.html?user=plantinformatics&repo=brioche&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>






<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Main</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Setting up a run</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#brioche-inputs">Brioche inputs:</a></li>
<li class="toctree-l2"><a class="reference internal" href="#parameters-file">Parameters file:</a></li>
<li class="toctree-l2"><a class="reference internal" href="#run-parameters-and-goals">Run parameters and goals</a></li>
<li class="toctree-l2"><a class="reference internal" href="#simple-runs">Simple runs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="results.html">Results</a></li>
<li class="toctree-l1"><a class="reference internal" href="citations.html">Citations</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">FAQ</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Use cases</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="usecases/usecase1_remapping.html">Usecase1: Remapping data across reference genomes</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecases/usecase2_insilico_genotyping.html">Usecase2: In silico genotyping of reference genomes</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecases/usecase3_redundancy_accuracy.html">Usecase3: Testing redundancy/accuracy in marker datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecases/usecase4_merge_datasets.html">Usecase4: Merge datasets</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="installation.html" title="previous chapter">Installation</a></li>
      <li>Next: <a href="results.html" title="next chapter">Results</a></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2026, James O'Dwyer, Gabriel Keeble-Gagnere, Kerrie L Forrest.
      
      |
      <a href="_sources/setting_up_a_run.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/plantinformatics/brioche" class="github">
        <img src="_static/github-banner.svg" alt="Fork me on GitHub" class="github"/>
    </a>
    

    
  </body>
</html>