<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Usecase1: Remapping data across reference genomes &#8212; Brioche 2.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=5ecbeea2" />
    <link rel="stylesheet" type="text/css" href="../_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=27fed22d" />
    <script src="../_static/documentation_options.js?v=51b770b3"></script>
    <script src="../_static/doctools.js?v=fd6eb6e6"></script>
    <script src="../_static/sphinx_highlight.js?v=6ffebe34"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Usecase2: In silico genotyping of reference genomes" href="usecase2_insilico_genotyping.html" />
    <link rel="prev" title="FAQ" href="../faq.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="usecase1-remapping-data-across-reference-genomes">
<h1>Usecase1: Remapping data across reference genomes<a class="headerlink" href="#usecase1-remapping-data-across-reference-genomes" title="Link to this heading">¶</a></h1>
<p>This usecase describes how to use Brioche to remap any population dataset against a new reference genome</p>
<p>The usecase assumes that a Brioche run has been completed. For more information about creating a Brioche run and results see</p>
<p>Setting up a run <a class="reference internal" href="../setting_up_a_run.html"><span class="doc">Setting up a run</span></a></p>
<p>Results of a run <a class="reference internal" href="../results.html"><span class="doc">Results folder structure</span></a></p>
<section id="requirements">
<h2>Requirements<a class="headerlink" href="#requirements" title="Link to this heading">¶</a></h2>
<ol class="arabic simple">
<li><p>Brioche to have been run and the files below are present</p></li>
</ol>
<p><em>‘*_Brioche_all_markers1to1stagingforvcf.csv’</em>
<em>‘*_marker_localdups_NULLS_counts.tsv’</em></p>
<p><em>‘*_priors_informed_strictmapping.tsv’</em></p>
<ol class="arabic" start="2">
<li><p>The genotype data file you have is in one of the following three forms:</p>
<ol class="loweralpha">
<li><p>Raw genotypes in the format below (tab-separated). <code class="docutils literal notranslate"><span class="pre">REF</span></code>/<code class="docutils literal notranslate"><span class="pre">ALT</span></code> are defined as ACGT/ACGT, samples are columns, markers are rows, genotypes are coded as <code class="docutils literal notranslate"><span class="pre">0</span></code>, <code class="docutils literal notranslate"><span class="pre">1</span></code>, <code class="docutils literal notranslate"><span class="pre">2</span></code>, <code class="docutils literal notranslate"><span class="pre">NC</span></code> as missing data, and <code class="docutils literal notranslate"><span class="pre">N</span></code> as null alleles.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>Name    REF  ALT  Sample1  Sample2  etc
Marker1 T    C    2        1
Marker2 T    G    0        N
Marker3 A    C    NC       1
</pre></div>
</div>
</li>
<li><p>A VCF file (any VCF-compatible format).</p></li>
<li><p>A DArTseq report (1-row format).</p></li>
</ol>
</li>
</ol>
</section>
<section id="runfile">
<h2>Runfile<a class="headerlink" href="#runfile" title="Link to this heading">¶</a></h2>
<p>To reanchor genotype data onto the new reference, a .sh script has been provided in the Additional_functions/ folder
Additional instructions are provided in the file but the main components to change are</p>
<ol class="arabic simple">
<li><p>The Slurm sbatch header commands e.g., #SBATCH –job-name=”example_anchoring”</p></li>
<li><p>the values of variable set in the section Variables to set in the script. There are 13 variables to set but most are changing the file paths and names of input/output files to the names generated in the specific brioche run</p></li>
</ol>
<a class="reference internal image-reference" href="../_images/Variables_anchoring_script.png"><img alt="Parameters file" class="align-center" src="../_images/Variables_anchoring_script.png" style="width: 300px;" />
</a>
<p>One variable the user can set which will change the result output files is the variable</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">donextstages=&quot;yes&quot;</span>
</pre></div>
</div>
<p>When set to “yes”, this will take the anchored VCF output file and</p>
<ol class="arabic simple">
<li><p>Create an additional VCF which has all markers which failed to map removed</p></li>
<li><p>Convert the chromosome names to numeric 1-n values and sort the new file</p></li>
</ol>
<p>This new file is ready for analysis in software sensitive to VCF settings and unmapped markers e.g., imputation</p>
</section>
<section id="outputs">
<h2>Outputs<a class="headerlink" href="#outputs" title="Link to this heading">¶</a></h2>
<p>The VCF file produced contains metadata from the brioche run as well as multiple specifics to Brioche outputs to assist with results interpretation
At the top of the VCF file we can see Brioche metadata included as well as some NCBI derived data for the reference genome. This reference data is queried from NCBI so requires the user to provide the Accession or genome name in the input variables when they run the script.</p>
<a class="reference internal image-reference" href="../_images/Anchored_vcf_output_headerCasestudy1.png"><img alt="Parameters file" class="align-center" src="../_images/Anchored_vcf_output_headerCasestudy1.png" style="width: 300px;" />
</a>
<p>In the body of the VCF file we can see a few of the additional values Brioche places into the VCF to assist with more informative downstream analysis</p>
<ol class="arabic simple">
<li><p>Brioche outputs two additional variables with each genotype</p></li>
</ol>
<p>Instead of a standard 1/1 0/0 etc, Brioche adds two additional fields NU and DU.</p>
<ol class="arabic simple">
<li><p>NU tracks the presence or absence of NULL allele states allowing for differentiation between missing data and true NULLs if the data type allows e.g., probe capture design</p></li>
<li><p>DU tracks the number of local duplications a marker hits to. For standard input genotype data this will always be . for unknown but when running insilico genotyping, added reference genomes will have counts for how many local duplications are detected.</p></li>
</ol>
<p>This allows for a greater understanding of why certain markers may have higher heterozygosity for example, with reference genomes having 2 or 3 copies of a region markers genotyping this region might be picking up multiple different mutations in your samples sequenced</p>
<p>Brioche VCF outputs as a result take the format 1/1:0:. or 0/0:0:. or ./.:1:. as just some examples.</p>
<p>2) Several additional format flags are given
e.g., MAPSTATUS=;PriorORIENT=;DUP=SingleCopy
This allows the user to easily see if the Mapping was unique through provided priors or not, whether the marker mapped positively or negatively to the reference genome, and whether there were local duplicates in the reference genome</p>
<a class="reference internal image-reference" href="../_images/Anchored_vcf_output_bodyCasestudy1.png"><img alt="Parameters file" class="align-center" src="../_images/Anchored_vcf_output_bodyCasestudy1.png" style="width: 300px;" />
</a>
<section id="other-usecases">
<h3>Other usecases<a class="headerlink" href="#other-usecases" title="Link to this heading">¶</a></h3>
<p>If you are interested in other usecases see.</p>
<p>2. If you are interested in extracting genotype calls from one or multiple reference genomes and adding them to your population genomics study go to
<a class="reference internal" href="usecase2_insilico_genotyping.html"><span class="doc">Usecase2: In silico genotyping of reference genomes</span></a></p>
<p>3. If you are interested in determining whether an existing marker dataset might be amplifying redundant regions under varied settings go to
<a class="reference internal" href="usecase3_redundancy_accuracy.html"><span class="doc">Usecase3: Testing redundancy/accuracy in marker datasets</span></a></p>
<p>4. If you are interested in the creation of custom marker datasets from existing analyses and testing the likely redundancy of newly designed markers against a wide range of reference genomes for a target species (similar process as 3.) go to
<a class="reference internal" href="usecase3_redundancy_accuracy.html"><span class="doc">Usecase3: Testing redundancy/accuracy in marker datasets</span></a></p>
<p>5. If you are interested in the mapping of multiple different datasets to a unified reference genome allowing for merging across shared loci and other downstream applications (e.g., imputations)
<a class="reference internal" href="usecase4_merge_datasets.html"><span class="doc">Usecase4: Merge datasets</span></a></p>
<p>otherwise, to return to the Introduction page go to <a class="reference internal" href="../introduction.html"><span class="doc">Introduction</span></a></p>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">Brioche</a></h1>






<p>
<iframe src="https://ghbtns.com/github-btn.html?user=plantinformatics&repo=brioche&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>






<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Main</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../setting_up_a_run.html">Setting up a run</a></li>
<li class="toctree-l1"><a class="reference internal" href="../results.html">Results</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citations.html">Citations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">FAQ</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Use cases</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usecase1: Remapping data across reference genomes</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#requirements">Requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="#runfile">Runfile</a></li>
<li class="toctree-l2"><a class="reference internal" href="#outputs">Outputs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="usecase2_insilico_genotyping.html">Usecase2: In silico genotyping of reference genomes</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecase3_redundancy_accuracy.html">Usecase3: Testing redundancy/accuracy in marker datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecase4_merge_datasets.html">Usecase4: Merge datasets</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
      <li>Previous: <a href="../faq.html" title="previous chapter">FAQ</a></li>
      <li>Next: <a href="usecase2_insilico_genotyping.html" title="next chapter">Usecase2: In silico genotyping of reference genomes</a></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2026, James O'Dwyer, Gabriel Keeble-Gagnere, Kerrie L Forrest.
      
      |
      <a href="../_sources/usecases/usecase1_remapping.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/plantinformatics/brioche" class="github">
        <img src="../_static/github-banner.svg" alt="Fork me on GitHub" class="github"/>
    </a>
    

    
  </body>
</html>