<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Usecase3: Testing redundancy/accuracy in marker datasets &#8212; Brioche 2.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=5ecbeea2" />
    <link rel="stylesheet" type="text/css" href="../_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=27fed22d" />
    <script src="../_static/documentation_options.js?v=51b770b3"></script>
    <script src="../_static/doctools.js?v=fd6eb6e6"></script>
    <script src="../_static/sphinx_highlight.js?v=6ffebe34"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Usecase4: Merge datasets" href="usecase4_merge_datasets.html" />
    <link rel="prev" title="Usecase2: In silico genotyping of reference genomes" href="usecase2_insilico_genotyping.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="usecase3-testing-redundancy-accuracy-in-marker-datasets">
<h1>Usecase3: Testing redundancy/accuracy in marker datasets<a class="headerlink" href="#usecase3-testing-redundancy-accuracy-in-marker-datasets" title="Link to this heading">Â¶</a></h1>
<p>This usecase describes how to use Brioche to test the redundancy/ accuracy of either existing or while designing marker datasets</p>
<p>Requirements</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>
1. Brioche targets file 

2. Reference genome FASTA/FASTAs

3. An updated params.config file ready to run Brioche


Runfile
~~~~~~~

To test for the redundancy/accuracy of markers in different settings a simple loop script is provided at Additional_functions/Run_brioche_optimise.sh
This file will allow the user to set multiple values for four variables to test as combinations to investigate how the proportion of uniquely mapping markers changes

Additional instructions are provided in the script but the main components to change are

1) The Slurm sbatch header commands e.g., #SBATCH --job-name=&quot;example_runbriocheoptimise&quot;

2) The different variable settings 
	a) Brioche specific settings (lines 34-63, 9 settings)
	b) Update the method for calling the required software (Nextflow, conda, and git) (Lines 64-71) currently written as module load 
	c) Thresholds to test (lines 73-79, 4 variable types) 




Output
~~~~~~

Brioche will be run for each combination of thresholds e.g., if identity_thresholds coverage_thresholds, maxhits, and wordsize 
all have 3 values each 81 (3^4) runs of brioche will be done so don&#39;t add too many at once to test.   

Brioche outputs will be placed in the chosen output folder with standard Brioche results but with
the name of each folder inlcuding the variables run in it e.g., Brioche_run_ident90cov70wordsize13


Extensions
~~~~~~~~~~

If the user is interested in identifying the most strictly mapped markers to design future markers from, this script allows for exploring what are the most strict parameters for a given marker dataset

After identifying strict settings, the user can then anchor the results as described in usecase 1 {doc}`Usecase1: Remapping data across reference genomes &lt;usecases/usecase1_remapping&gt;`
take the VCF of only mapped reads and extract markers from it using the Additional_functions/run_convert_vcf_to_brioche_input.sh
and then run Brioche on a new reference genome iteratively restricting the dataset to only the best mapping markers across all reference genomes



Other usecases
==============

If you are interested in other usecases see.

1. If you are interested in easy remapping of markers across any distinct reference genome and the reanchoring of genotypes to the new reference genome 
{doc}`Usecase1: Remapping data across reference genomes &lt;usecases/usecase1_remapping&gt;`

2. If you are interested in extracting genotype calls from one or multiple reference genomes and adding them to your population genomics study go to 
{doc}`Usecase2: In silico genotyping of reference genomes &lt;usecases/usecase2_insilico_genotyping&gt;`

5. If you are interested in the mapping of multiple different datasets to a unified reference genome allowing for merging across shared loci and other downstream applications (e.g., imputations)
{doc}`Usecase4: Merge datasets &lt;usecases/usecase4_merge_datasets&gt;`


otherwise, to return to the Introduction page go to {doc}`Introduction &lt;../Introduction&gt;`


</pre></div>
</div>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">Brioche</a></h1>






<p>
<iframe src="https://ghbtns.com/github-btn.html?user=plantinformatics&repo=brioche&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>






<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Main</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Brioche overview:</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html#key-features">Key features:</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html#key-use-cases">Key use cases:</a></li>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html#workflow-summary">Workflow summary:</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html#downloading-brioche">Downloading brioche:</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html#software-dependencies">Software dependencies:</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html#file-layout">File layout:</a></li>
<li class="toctree-l1"><a class="reference internal" href="../setting_up_a_run.html">Setting up a run</a></li>
<li class="toctree-l1"><a class="reference internal" href="../setting_up_a_run.html#brioche-inputs">Brioche inputs:</a></li>
<li class="toctree-l1"><a class="reference internal" href="../setting_up_a_run.html#parameters-file">Parameters file:</a></li>
<li class="toctree-l1"><a class="reference internal" href="../results.html">Results folder structure</a></li>
<li class="toctree-l1"><a class="reference internal" href="../results.html#summary-report">Summary report</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citations.html">Citations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">FAQ</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Use cases</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="usecase1_remapping.html">Usecase1: Remapping data across reference genomes</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecase2_insilico_genotyping.html">Usecase2: In silico genotyping of reference genomes</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usecase3: Testing redundancy/accuracy in marker datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecase4_merge_datasets.html">Usecase4: Merge datasets</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
      <li>Previous: <a href="usecase2_insilico_genotyping.html" title="previous chapter">Usecase2: In silico genotyping of reference genomes</a></li>
      <li>Next: <a href="usecase4_merge_datasets.html" title="next chapter">Usecase4: Merge datasets</a></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2026, James O'Dwyer, Gabriel Keeble-Gagnere, Kerrie L Forrest.
      
      |
      <a href="../_sources/usecases/usecase3_redundancy_accuracy.md.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/plantinformatics/brioche" class="github">
        <img src="../_static/github-banner.svg" alt="Fork me on GitHub" class="github"/>
    </a>
    

    
  </body>
</html>