<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Usecase4: Merge datasets &#8212; Brioche 2.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=5ecbeea2" />
    <link rel="stylesheet" type="text/css" href="../_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=27fed22d" />
    <script src="../_static/documentation_options.js?v=51b770b3"></script>
    <script src="../_static/doctools.js?v=fd6eb6e6"></script>
    <script src="../_static/sphinx_highlight.js?v=6ffebe34"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="Usecase3: Testing redundancy/accuracy in marker datasets" href="usecase3_redundancy_accuracy.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="usecase4-merge-datasets">
<h1>Usecase4: Merge datasets<a class="headerlink" href="#usecase4-merge-datasets" title="Link to this heading">¶</a></h1>
<p>This usecase describes how to use Brioche to map multiple datasets for merging.</p>
<p>This usecase is an extension of Usecase1 but with distinct defined goals and as such will only briefly touch on how to set up</p>
<section id="requirements">
<h2>Requirements<a class="headerlink" href="#requirements" title="Link to this heading">¶</a></h2>
<p>Per dataset of interest:</p>
<p>1. Brioche to have been run and the files below are present
<em>‘*_Brioche_all_markers1to1stagingforvcf.csv’</em></p>
<p><em>‘*_marker_localdups_NULLS_counts.tsv’</em></p>
<p><em>‘*_priors_informed_strictmapping.tsv’</em></p>
<ol class="arabic" start="2">
<li><p>The genotype data file you have is in one of the following three forms:</p>
<ol class="loweralpha">
<li><p>Raw genotypes in the format below (tab-separated). <code class="docutils literal notranslate"><span class="pre">REF</span></code>/<code class="docutils literal notranslate"><span class="pre">ALT</span></code> are defined as ACGT/ACGT, samples are columns, markers are rows, genotypes are coded as <code class="docutils literal notranslate"><span class="pre">0</span></code>, <code class="docutils literal notranslate"><span class="pre">1</span></code>, <code class="docutils literal notranslate"><span class="pre">2</span></code>, <code class="docutils literal notranslate"><span class="pre">NC</span></code> as missing data, and <code class="docutils literal notranslate"><span class="pre">N</span></code> as null alleles.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>Name    REF  ALT  Sample1  Sample2  etc
Marker1 T    C    2        1
Marker2 T    G    0        N
Marker3 A    C    NC       1
</pre></div>
</div>
</li>
<li><p>A VCF file (any VCF-compatible format).</p></li>
<li><p>A DArTseq report (1-row format).</p></li>
</ol>
</li>
</ol>
</section>
<section id="runfile">
<h2>Runfile<a class="headerlink" href="#runfile" title="Link to this heading">¶</a></h2>
<p>Per dataset of interest: to reanchor genotype data onto the new reference, a .sh script has been provided in the Additional_functions/ folder
Additional instructions are provided in the file but the main components to change are</p>
<ol class="arabic simple">
<li><p>The Slurm sbatch header commands e.g., #SBATCH –job-name=”example_anchoring”</p></li>
<li><p>the values of variable set in the section Variables to set in the script. There are 13 variables to set but most are changing the file paths and names of input/output files to the names generated in the specific brioche run</p></li>
</ol>
<a class="reference internal image-reference" href="../_images/Variables_anchoring_script.png"><img alt="variables anchoring script" class="align-center" src="../_images/Variables_anchoring_script.png" style="width: 300px;" />
</a>
<p>the below setting is required</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">donextstages=&quot;yes&quot;</span>
</pre></div>
</div>
</section>
<section id="outputs">
<h2>Outputs<a class="headerlink" href="#outputs" title="Link to this heading">¶</a></h2>
<p>Brioche will produce a number of outputs similar to casestudy2 <a class="reference internal" href="usecase2_insilico_genotyping.html"><span class="doc">Usecase2: In silico genotyping of reference genomes</span></a></p>
<p>These are</p>
<ol class="arabic simple">
<li><p>a VCF output with all markers</p></li>
<li><p>A VCF with only mapped markers</p></li>
<li><p>A VCF with mapped markers which have numeric chromosomes</p></li>
<li><p>A VCF with mapped markers which have numeric chromosomes which are sorted</p></li>
</ol>
<p>For each of the analysed datasets the user can take</p>
<ol class="arabic simple" start="4">
<li><p>A VCF with mapped markers which have numeric chromosomes which are sorted</p></li>
</ol>
<p>and if there is sufficient overlap in markers the user can directly merge the new datasets</p>
<p>e.g., through first activating the brioche-vcf conda environment and then running bcftools</p>
<p><em>‘union_two_datasets.vcf.gz’</em> will have all markers and all samples across both datasets</p>
<p><em>‘isec_merged_results’</em> will be a folder with 3 vcf files, one for unique markers in dataset1, one for unique in dataset2, and one for shared markers</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">conda activate brioche-vcf</span>
<span class="go">gzip dataset1.vcf</span>
<span class="go">gzip dataset2.vcf</span>
<span class="go">bcftools index -c dataset1.vcf.gz</span>
<span class="go">bcftools index -c dataset2.vcf.gz</span>
<span class="go">bcftools merge \</span>
<span class="go"> -m none \</span>
<span class="go"> -Oz \</span>
<span class="go"> -o merged_union_two_datasets.vcf.gz \</span>
<span class="go"> dataset1.vcf.gz dataset2.gz</span>
<span class="go">bcftools isec -p isec_merged_results -c all dataset1.vcf.gz dataset2.vcf.gz</span>
</pre></div>
</div>
<p>Alternatively if there is still poor overlap each dataset can be run through imputation and the imputed datasets can be merged for greater overlap</p>
<section id="other-usecases">
<h3>Other usecases<a class="headerlink" href="#other-usecases" title="Link to this heading">¶</a></h3>
<p>If you are interested in other usecases see.</p>
<p>1. If you are interested in easy remapping of markers across any distinct reference genome and the reanchoring of genotypes to the new reference genome
<a class="reference internal" href="usecase1_remapping.html"><span class="doc">Usecase1: Remapping data across reference genomes</span></a></p>
<p>2. If you are interested in extracting genotype calls from one or multiple reference genomes and adding them to your population genomics study go to
<a class="reference internal" href="usecase2_insilico_genotyping.html"><span class="doc">Usecase2: In silico genotyping of reference genomes</span></a></p>
<p>3. If you are interested in determining whether an existing marker dataset might be amplifying redundant regions under varied settings go to
<a class="reference internal" href="usecase3_redundancy_accuracy.html"><span class="doc">Usecase3: Testing redundancy/accuracy in marker datasets</span></a></p>
<p>4. If you are interested in the creation of custom marker datasets from existing analyses and testing the likely redundancy of newly designed markers against a wide range of reference genomes for a target species (similar process as 3.) go to
<a class="reference internal" href="usecase3_redundancy_accuracy.html"><span class="doc">Usecase3: Testing redundancy/accuracy in marker datasets</span></a></p>
<p>otherwise, to return to the Introduction page go to <a class="reference internal" href="../introduction.html"><span class="doc">Introduction</span></a></p>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">Brioche</a></h1>






<p>
<iframe src="https://ghbtns.com/github-btn.html?user=plantinformatics&repo=brioche&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>






<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Main</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../setting_up_a_run.html">Setting up a run</a></li>
<li class="toctree-l1"><a class="reference internal" href="../results.html">Results</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citations.html">Citations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">FAQ</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Use cases</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="usecase1_remapping.html">Usecase1: Remapping data across reference genomes</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecase2_insilico_genotyping.html">Usecase2: In silico genotyping of reference genomes</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecase3_redundancy_accuracy.html">Usecase3: Testing redundancy/accuracy in marker datasets</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usecase4: Merge datasets</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#requirements">Requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="#runfile">Runfile</a></li>
<li class="toctree-l2"><a class="reference internal" href="#outputs">Outputs</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
      <li>Previous: <a href="usecase3_redundancy_accuracy.html" title="previous chapter">Usecase3: Testing redundancy/accuracy in marker datasets</a></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2026, James O'Dwyer, Gabriel Keeble-Gagnere, Kerrie L Forrest.
      
      |
      <a href="../_sources/usecases/usecase4_merge_datasets.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/plantinformatics/brioche" class="github">
        <img src="../_static/github-banner.svg" alt="Fork me on GitHub" class="github"/>
    </a>
    

    
  </body>
</html>