<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Usecase2: In silico genotyping of reference genomes &#8212; Brioche 2.0.0 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=5ecbeea2" />
    <link rel="stylesheet" type="text/css" href="../_static/basic.css?v=b08954a9" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css?v=27fed22d" />
    <script src="../_static/documentation_options.js?v=51b770b3"></script>
    <script src="../_static/doctools.js?v=fd6eb6e6"></script>
    <script src="../_static/sphinx_highlight.js?v=6ffebe34"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Usecase3: Testing redundancy/accuracy in marker datasets" href="usecase3_redundancy_accuracy.html" />
    <link rel="prev" title="Usecase1: Remapping data across reference genomes" href="usecase1_remapping.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="usecase2-in-silico-genotyping-of-reference-genomes">
<h1>Usecase2: In silico genotyping of reference genomes<a class="headerlink" href="#usecase2-in-silico-genotyping-of-reference-genomes" title="Link to this heading">¶</a></h1>
<p>This usecase describes how Brioche can be used to extract equivalent genotypes from any number of desired reference genomes to add them to sample sets for additional analyses.</p>
<p>This usecase assumes that the user is familiar with the basic structure of the Brioche results and what the anchored genotypes VCF file looks like but does not require a brioche run to be complete yet</p>
<p>Generate in silico genotypes by mapping marker sequences to a reference genome.</p>
<section id="requirements">
<h2>Requirements<a class="headerlink" href="#requirements" title="Link to this heading">¶</a></h2>
<ol class="arabic">
<li><p>Brioche targets file</p></li>
<li><p>The genotype data file you have is in one of the following three forms:</p>
<ol class="loweralpha">
<li><p>Raw genotypes in the format below (tab-separated). <code class="docutils literal notranslate"><span class="pre">REF</span></code>/<code class="docutils literal notranslate"><span class="pre">ALT</span></code> are defined as ACGT/ACGT, samples are columns, markers are rows, genotypes are coded as <code class="docutils literal notranslate"><span class="pre">0</span></code>, <code class="docutils literal notranslate"><span class="pre">1</span></code>, <code class="docutils literal notranslate"><span class="pre">2</span></code>, <code class="docutils literal notranslate"><span class="pre">NC</span></code> as missing data, and <code class="docutils literal notranslate"><span class="pre">N</span></code> as null alleles.</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>Name    REF  ALT  Sample1  Sample2  etc
Marker1 T    C    2        1
Marker2 T    G    0        N
Marker3 A    C    NC       1
</pre></div>
</div>
</li>
<li><p>A VCF file (any VCF-compatible format).</p></li>
<li><p>A DArTseq report (1-row format).</p></li>
</ol>
</li>
<li><p>Reference genome FASTA</p></li>
<li><p>An updated <code class="docutils literal notranslate"><span class="pre">params.config</span></code> file ready to run Brioche</p></li>
<li><p>A list of full paths/filenames to reference genomes to in silico genotype (see <code class="docutils literal notranslate"><span class="pre">example_files/genomes_insilico.txt</span></code>)</p></li>
<li><p>Installation of Nextflow, conda, git, and R</p></li>
</ol>
</section>
<section id="runfile">
<h2>Runfile<a class="headerlink" href="#runfile" title="Link to this heading">¶</a></h2>
<p>To run insilico genotyping, a .sh script file is provided in Additional_functions/Run_insilico_add.sh
Additional instructions are provided in the script but the main components to change are.</p>
<ol class="arabic simple">
<li><p>The Slurm <code class="docutils literal notranslate"><span class="pre">sbatch</span></code> header commands (e.g., <code class="docutils literal notranslate"><span class="pre">#SBATCH</span> <span class="pre">--job-name=&quot;example_insilicorunning&quot;</span></code>)</p></li>
<li><p>Four different variable settings:</p>
<ol class="loweralpha simple">
<li><p>Brioche settings (lines 33-58, 7 settings). These settings require the user to update the paths to where Brioche was downloaded and where the targets file is located.</p></li>
<li><p>Output folder name (line 62, 1 setting). This setting describes where all in silico results should be saved.</p></li>
<li><p>Anchoring settings (lines 66-82, 6 settings). These settings point to the anchoring script and define the initial genotype format (e.g., <code class="docutils literal notranslate"><span class="pre">isvcf</span></code> or <code class="docutils literal notranslate"><span class="pre">isdart</span></code> or <code class="docutils literal notranslate"><span class="pre">issnp</span></code>).</p></li>
<li><p>Final Brioche settings (lines 84-95, 4 settings). These settings specify the final reference genome and what final VCF the genotypes will be anchored against.</p></li>
<li><p>Update the method for calling the required software (Nextflow, conda, git, and R) (lines 97-105). These are currently written as <code class="docutils literal notranslate"><span class="pre">module</span> <span class="pre">load</span></code> statements.</p></li>
</ol>
</li>
</ol>
<a class="reference internal image-reference" href="../_images/Brioche_insilico_variables.png"><img alt="Insilico variables" class="align-center" src="../_images/Brioche_insilico_variables.png" style="width: 300px;" />
</a>
</section>
<section id="outputs">
<h2>Outputs<a class="headerlink" href="#outputs" title="Link to this heading">¶</a></h2>
<p>The results of insilico genotyping are output into three separate folders in the chosen output directory</p>
<ol class="arabic simple">
<li><p>intermediate_brioche/ where all Brioche runs are stored e.g.,  intermediate_brioche/genomename/brioche-results (one per reference genome)</p></li>
<li><p>anchoring/where all genomes are iteratively anchored into a single VCF</p></li>
<li><p>final/ where the final brioche run and final anchored results are stored.</p></li>
</ol>
<p>For a detailed breakdown of 1) see <a class="reference internal" href="../results.html"><span class="doc">Results folder structure</span></a></p>
<ol class="arabic simple" start="2">
<li><p>contains intermediate data and is not directly relevant for downstream analysis after Brioche</p></li>
<li><p>In the final folder the Brioche results and reports can be viewed but the primary result is in /anchored_results</p></li>
</ol>
<a class="reference internal image-reference" href="../_images/Brioche_insilico_anchored_results_folder.png"><img alt="insilico results folder" class="align-center" src="../_images/Brioche_insilico_anchored_results_folder.png" style="width: 300px;" />
</a>
<p>In this folder there will be several output files of interest</p>
<ol class="arabic simple">
<li><p><em>‘Building_final_mapped_against_IWGSC_RefSeq_v2.1_refs.vcf’</em> A VCF of all markers containing the genotypes of the reference genomes (plus the two leading samples of the provided genotypes file)</p></li>
<li><p><em>‘Building_final_mapped_against_IWGSC_RefSeq_v2.1_seqs.vcf’</em> A VCF of all markers containing the genotypes of all samples from the genotypes file</p></li>
<li><p><em>‘Building_final_mapped_against_IWGSC_RefSeq_v2.1.vcf’</em> A VCF containing both the sequenced samples and the insilico genotyped references with markers which failed to map uniquely removed</p></li>
<li><p><em>‘Building_final_mapped_against_IWGSC_RefSeq_v2.1.vcf.gz’</em> A gzip of the above file</p></li>
<li><p><em>‘Building_final_mapped_against_IWGSC_RefSeq_v2.1.vcf_numericchroms.vcf’</em> A VCF with the name of chromosomes converted to numeric format</p></li>
</ol>
<p>Chromosomes are converted to numeric format either through the user providing a chromosome matching file (see example_files/example_chrommatching_for_converting_vcf_chromnames_to_numeric.tsv)
If no file is provided the chroms will be sorted through pattern matching names to look for numeric values
If chromosomes lack numeric patterns, chromosomes will be renamed numerically from 1-n based on their order in the reference genome used</p>
<ol class="arabic simple">
<li><p><em>‘Building_final_mapped_against_IWGSC_RefSeq_v2.1.vcf_failedtomap.vcf’</em> A VCF of all markers which failed to map uniquely</p></li>
<li><p><em>‘unmapped_markers.tsv’</em> A one column tsv list of markers which failed to map uniquely</p></li>
<li><p><em>‘Building_final_mapped_against_IWGSC_RefSeq_v2.1.vcf_sorted.vcf’</em> A VCF copy of  ‘Building_final_mapped_against_IWGSC_RefSeq_v2.1.vcf_numericchroms.vcf’ which has been sorted numerically by chromosome and position</p></li>
</ol>
<p>Looking deeper into one of the markers we can see some of the novel results insilico genotyping can achieve. Below is a line containing only reference genome genotypes from an insilico run
For this marker on chromosome 20 it can be seen that form the majority of genomes this marker has 2+ local duplications (the end value per genotype 0/0:0:<strong>2</strong>) This shows that the marker in question could represent a large variable duplication site and may be relevant if linked to genes of known function
Alternatively, if the marker is in the middle of a region of non coding DNA this marker may be performing more poorly during sequencing because of the number of duplication sites)</p>
<p>‘20     744202465       Marker10        G       A       100     PASS    MAPSTATUS=Unique_mapping;PriorORIENT=plus;DUP=LocallyDuplicatedRegion;MAF=0;AN=25202;AC=0       GT:NU:DU        0/0:0:2 0/0:0:3 ./.:1:. 0/0:0:0 0/0:0:3 0/0:0:3 0/0:0:3 0/0:0:3 ./.:1:. 0/0:0:3 0/0:0:0’</p>
<p>Now the file ‘Building_final_mapped_against_IWGSC_RefSeq_v2.1.vcf_sorted.vcf’ can be taken and put into clustering analyses like PCAs to see where samples sit relative to known genomes etc.</p>
<section id="other-usecases">
<h3>Other usecases<a class="headerlink" href="#other-usecases" title="Link to this heading">¶</a></h3>
<p>If you are interested in other usecases see.</p>
<p>1. If you are interested in easy remapping of markers across any distinct reference genome and the reanchoring of genotypes to the new reference genome
<a class="reference internal" href="usecase1_remapping.html"><span class="doc">Usecase1: Remapping data across reference genomes</span></a></p>
<p>3. If you are interested in determining whether an existing marker dataset might be amplifying redundant regions under varied settings go to
<a class="reference internal" href="usecase3_redundancy_accuracy.html"><span class="doc">Usecase3: Testing redundancy/accuracy in marker datasets</span></a></p>
<p>4. If you are interested in the creation of custom marker datasets from existing analyses and testing the likely redundancy of newly designed markers against a wide range of reference genomes for a target species (similar process as 3.) go to
<a class="reference internal" href="usecase3_redundancy_accuracy.html"><span class="doc">Usecase3: Testing redundancy/accuracy in marker datasets</span></a></p>
<p>5. If you are interested in the mapping of multiple different datasets to a unified reference genome allowing for merging across shared loci and other downstream applications (e.g., imputations)
<a class="reference internal" href="usecase4_merge_datasets.html"><span class="doc">Usecase4: Merge datasets</span></a></p>
<p>otherwise, to return to the Introduction page go to <a class="reference internal" href="../introduction.html"><span class="doc">Introduction</span></a></p>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">Brioche</a></h1>






<p>
<iframe src="https://ghbtns.com/github-btn.html?user=plantinformatics&repo=brioche&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>






<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Main</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../setting_up_a_run.html">Setting up a run</a></li>
<li class="toctree-l1"><a class="reference internal" href="../results.html">Results</a></li>
<li class="toctree-l1"><a class="reference internal" href="../citations.html">Citations</a></li>
<li class="toctree-l1"><a class="reference internal" href="../faq.html">FAQ</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">Use cases</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="usecase1_remapping.html">Usecase1: Remapping data across reference genomes</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usecase2: In silico genotyping of reference genomes</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#requirements">Requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="#runfile">Runfile</a></li>
<li class="toctree-l2"><a class="reference internal" href="#outputs">Outputs</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="usecase3_redundancy_accuracy.html">Usecase3: Testing redundancy/accuracy in marker datasets</a></li>
<li class="toctree-l1"><a class="reference internal" href="usecase4_merge_datasets.html">Usecase4: Merge datasets</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
      <li>Previous: <a href="usecase1_remapping.html" title="previous chapter">Usecase1: Remapping data across reference genomes</a></li>
      <li>Next: <a href="usecase3_redundancy_accuracy.html" title="next chapter">Usecase3: Testing redundancy/accuracy in marker datasets</a></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2026, James O'Dwyer, Gabriel Keeble-Gagnere, Kerrie L Forrest.
      
      |
      <a href="../_sources/usecases/usecase2_insilico_genotyping.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/plantinformatics/brioche" class="github">
        <img src="../_static/github-banner.svg" alt="Fork me on GitHub" class="github"/>
    </a>
    

    
  </body>
</html>